src/Mod/Part/Gui/ViewProviderGridExtension.cpp

```diff
@@ -0,0 +1,479 @@
+    int GridSizePixelThreshold = 15;
+    int GridNumberSubdivision = 10;
+    int GridLinePattern = 0x0f0f;
+    int GridDivLinePattern = 0xffff;
+    int GridLineWidth = 1;
+    int GridDivLineWidth = 2;
+    unsigned int GridLineColor;
+    unsigned int GridDivLineColor;
+
+    // VP Independent parameters through observer
+    int unitsUserSchema = 0;
+
+    /** Observer for parameter group. */
+    void OnChange(Base::Subject<const char*> &rCaller, const char * sReason) override;
+
+private:
+    double getGridSize(const Gui::View3DInventorViewer* viewer);
+    SoSeparator* createGrid(bool cameraUpdate = false);
+    SoSeparator* createGridPart(double Step, int numberSubdiv, bool divLines, bool subDivLines, int pattern, SoBaseColor* color, int lineWidth = 1);
+
+    void createEditModeInventorNodes();
+
+    SbVec3f camCenterOnSketch;
+    float camMaxDimension;
+
+private:
+    ViewProviderGridExtension * vp;
+
+    bool enabled = false;
+
+    // scenograph
+    SoSeparator * GridRoot;
+};
+
+} // namespace PartGui
+
+
+GridExtensionP::GridExtensionP(ViewProviderGridExtension * vp):
+    camCenterOnSketch(SbVec3f(0., 0., 0.)),
+    camMaxDimension(200.),
+    vp(vp),
+    GridRoot(nullptr)
+{
+    SbColor lineCol(0.7f, 0.7f, 0.7f);
+    GridLineColor = lineCol.getPackedValue();
+    GridDivLineColor = GridLineColor;
+
+    createEditModeInventorNodes();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Units");
+    unitsUserSchema = hGrp->GetInt("UserSchema", 0); //2 3 5 7 are imperial schemas. 2 3 inches, 5 7 feet
+    hGrp->Attach(this);
+}
+
+GridExtensionP::~GridExtensionP()
+{
+    Gui::coinRemoveAllChildren(GridRoot);
+    GridRoot->unref();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Units");
+    hGrp->Detach(this);
+}
+
+double GridExtensionP::getGridSize(const Gui::View3DInventorViewer* viewer)
+{
+    short pixelWidth = -1;
+    short pixelHeight = -1;
+    viewer->getViewportRegion().getViewportSizePixels().getValue(pixelWidth, pixelHeight);
+    if (pixelWidth < 0 || pixelHeight < 0)
+        return vp->GridSize.getValue();
+
+    int numberOfLines = std::max(pixelWidth, pixelHeight) / GridSizePixelThreshold;
+
+    double unitMultiplier = (unitsUserSchema == 2 || unitsUserSchema == 3) ? 25.4 : (unitsUserSchema == 5 || unitsUserSchema == 7) ? 304.8 : 1;
+
+    double newGridSize = unitMultiplier * pow(GridNumberSubdivision, 1 + floor(log(camMaxDimension / unitMultiplier / numberOfLines) / log(GridNumberSubdivision)));
+
+    //cap the grid size
+    newGridSize = std::max(newGridSize, 0.000001);
+    newGridSize = std::min(newGridSize, 10000000.0);
+
+    if (newGridSize != vp->GridSize.getValue()) // protect from recursive calls
+        vp->GridSize.setValue(newGridSize); //grid size must be set for grid snap. But we need to block it from calling createGrid.
```

Comment:
I find it feasible that the grid is asked to compute itself with the same parameters. Not necessarily because of a recursive call, but because "createGrid" is called and the camera center moved (for example), in that case in the grid mode auto is active, the "computeGrid" function is called. Unconditionally setting the grid to the same value it might have, would generate a property change, when it appears to me none if necessary.

This check avoids this unnecessary update. The comment could certainly be updated.

Do you agree or see it otherwise?

Reply:
I trust you --not much time to have deep checks :)--. And anyway it doesn't hurt. ;)